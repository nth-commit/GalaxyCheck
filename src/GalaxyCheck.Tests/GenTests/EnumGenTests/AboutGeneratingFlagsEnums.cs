using FluentAssertions;
using GalaxyCheck;
using NebulaCheck;
using Snapshooter;
using Snapshooter.Xunit;
using System;
using System.Linq;
using Xunit;
using static Tests.V2.DomainGenAttributes;

namespace Tests.V2.GenTests.EnumGenTests
{
    public class AboutGeneratingFlagsEnums
    {
        [Flags]
        private enum AllowedOperations
        {
            Create = 1 << 0,

            Read = 1 << 1,

            Update = 1 << 2,

            Delete = 1 << 3
        }

        [Property]
        public void ItShrinksTowardsTheSmallestValue([Seed] int seed)
        {
            var gen = GalaxyCheck.Gen.Enum<AllowedOperations>();

            var minimum = gen.Minimum(seed: seed);

            minimum.Should().Be(AllowedOperations.Create);
        }

        [Fact]
        public void Snapshots()
        {
            // Should demonstrate the combinatorial factor when generating a flags enum

            var seeds = Enumerable.Range(0, 3);

            var gen = GalaxyCheck.Gen.Enum<AllowedOperations>();

            foreach (var seed in seeds)
            {
                var sample = gen
                    .Advanced.SampleOneExampleSpace(seed: seed, size: 75)
                    .Take(500)
                    .Render(x => x.ToString());

                var nameExtension = string.Join("_", new[]
                {
                    $"Seed_{seed}"
                });

                Snapshot.Match(sample, SnapshotNameExtension.Create(nameExtension));
            }
        }
    }
}
