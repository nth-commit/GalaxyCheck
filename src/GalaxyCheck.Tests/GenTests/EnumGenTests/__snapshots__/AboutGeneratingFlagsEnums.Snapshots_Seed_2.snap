Create, Update, Delete (Id = 1550704683, Distance = 233.3333...)
├> Create, Read, Delete (Id = -954692729, Distance = 200.0000)
|  ├> Delete (Id = -1521134292, Distance = 100)
|  ├> Read, Delete (Id = -149127010, Distance = 166.6667...)
|  |  └> Read (Id = -1521134294, Distance = 33.3333...)
|  ├> Create, Read (Id = -1401825717, Distance = 66.6667...)
|  |  ├> Read (Id = -1521134294, Distance = 33.3333...)
|  |  └> Create (Id = -1521134295, Distance = 0)
|  └> Create, Delete (Id = -149127011, Distance = 133.3333...)
|     └> Create (Id = -1521134295, Distance = 0)
├> Delete (Id = -1521134292, Distance = 100)
├> Update, Delete (Id = -149127009, Distance = 200.0000)
|  ├> Update (Id = -1521134293, Distance = 66.6667...)
|  |  └> Read (Id = -1521134294, Distance = 33.3333...)
|  └> Read, Delete (Id = -149127010, Distance = 166.6667...)
|     └> Read (Id = -1521134294, Distance = 33.3333...)
├> Create, Update (Id = 1372007284, Distance = 100.0000)
|  ├> Update (Id = -1521134293, Distance = 66.6667...)
|  |  └> Read (Id = -1521134294, Distance = 33.3333...)
|  ├> Create (Id = -1521134295, Distance = 0)
|  └> Create, Read (Id = -1401825717, Distance = 66.6667...)
|     └> Read (Id = -1521134294, Distance = 33.3333...)
└> Create, Delete (Id = -149127011, Distance = 133.3333...)
   └> Create (Id = -1521134295, Distance = 0)
